File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Core/Tva/Schedule.vb
' 08-01-06

Option Strict On

Imports System.Text.RegularExpressions
Imports Tva.Text.Common

Public Class Schedule

#Region " Private Schedule Element Class "

    Private Class Element

        Private m_text As String
        Private m_elementType As ElementType
        Private m_values As List(Of Integer)

        Public Enum ElementType As Integer
            Minute
            Hour
            Day
            Month
            DayOfWeek
        End Enum

        Public Sub New(ByVal text As String, ByVal type As ElementType)
            If IsValidElement(text, type) Then
                m_text = text
                m_elementType = type
            Else
                Throw New ArgumentException()
            End If
        End Sub

        Public ReadOnly Property Text() As String
            Get
                Return m_text
            End Get
        End Property

        Public ReadOnly Property Values() As List(Of Integer)
            Get
                Return m_values
            End Get
        End Property

        Public Function Matches(ByVal dateAndTime As System.DateTime) As Boolean

            Select Case m_elementType
                Case ElementType.Minute
                    Return m_values.Contains(dateAndTime.Minute)
                Case ElementType.Hour
                    Return m_values.Contains(dateAndTime.Hour)
                Case ElementType.Day
                    Return m_values.Contains(dateAndTime.Day)
                Case ElementType.Month
                    Return m_values.Contains(dateAndTime.Month)
                Case ElementType.DayOfWeek
                    Return m_values.Contains(Convert.ToInt32(dateAndTime.DayOfWeek))
            End Select

        End Function

        Private Function IsValidElement(ByVal element As String, ByVal elementType As ElementType) As Boolean

            Dim minValue As Integer = 0
            Dim maxvalue As Integer = 0
            Select Case elementType
                Case Schedule.Element.ElementType.Minute
                    maxvalue = 59
                Case Schedule.Element.ElementType.Hour
                    maxvalue = 23
                Case Schedule.Element.ElementType.Day
                    minValue = 1
                    maxvalue = 31
                Case Schedule.Element.ElementType.Month
                    minValue = 1
                    maxvalue = 12
                Case Schedule.Element.ElementType.DayOfWeek
                    maxvalue = 6
            End Select

            m_values = New List(Of Integer)()
            If Regex.Match(element, "^(\*){1}$").Success Then
                ' ^(\*){1}$             Matches: *
                PopulateValues(minValue, maxvalue, 1)

                Return True
            ElseIf Regex.Match(element, "^(\*/\d+){1}$").Success Then
                ' ^(\*/\d+){1}$         Matches: */[any digit]
                Dim interval As Integer = Convert.ToInt32(element.Split("/"c)(1))
                If interval > 0 AndAlso interval >= minValue AndAlso interval <= maxvalue Then
                    PopulateValues(minValue, maxvalue, interval)
                    Return True
                End If

                Return False
            ElseIf Regex.Match(element, "^(\d+\-\d+){1}$").Success Then
                ' ^(\d+\-\d+){1}$       Matches: [any digit]-[any digit]
                Dim range As String() = element.Split("-"c)
                Dim lowRange As Integer = Convert.ToInt32(range(0))
                Dim highRange As Integer = Convert.ToInt32(range(1))
                If lowRange < highRange AndAlso lowRange >= minValue AndAlso highRange <= maxvalue Then
                    PopulateValues(lowRange, highRange, 1)
                    Return True
                End If

                Return False
            ElseIf Regex.Match(element, "^((\d+,?)+){1}$").Success Then
                ' ^((\d+,?)+){1}$       Matches: [any digit] AND [any digit], ..., [any digit]
                For Each value As Integer In element.Split(","c)
                    If Not (value >= minValue AndAlso value <= maxvalue) Then
                        Return False
                    Else
                        If Not m_values.Contains(value) Then m_values.Add(value)
                    End If
                Next

                Return True
            End If

        End Function

        Private Sub PopulateValues(ByVal fromValue As Integer, ByVal toValue As Integer, ByVal stepValue As Integer)

            For i As Integer = fromValue To toValue Step stepValue
                m_values.Add(i)
            Next

        End Sub

    End Class

#End Region

    Private m_name As String
    Private m_minutes As Element
    Private m_hours As Element
    Private m_days As Element
    Private m_months As Element
    Private m_dayOfWeek As Element
    Private m_lastRunDateTime As System.DateTime

    Public Sub New(ByVal name As String)
        MyClass.New(name, "*")
    End Sub

    Public Sub New(ByVal name As String, ByVal minutes As String)
        MyClass.New(name, minutes, "*")
    End Sub

    Public Sub New(ByVal name As String, ByVal minutes As String, ByVal hours As String)
        MyClass.New(name, minutes, hours, "*")
    End Sub

    Public Sub New(ByVal name As String, ByVal minutes As String, ByVal hours As String, ByVal days As String)
        MyClass.New(name, minutes, hours, days, "*")
    End Sub

    Public Sub New(ByVal name As String, ByVal minutes As String, ByVal hours As String, ByVal days As String, _
                ByVal months As String)
        MyClass.New(name, minutes, hours, days, months, "*")
    End Sub
    Public Sub New(ByVal name As String, ByVal minutes As String, ByVal hours As String, ByVal days As String, _
            ByVal months As String, ByVal daysOfWeek As String)
        MyBase.New()
        Me.Name = name
        Me.Rule = minutes & " " & hours & " " & days & " " & months & " " & daysOfWeek
    End Sub

    ''' <summary>
    ''' Gets or sets the schedule name.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The schedule name.</returns>
    Public Property Name() As String
        Get
            Return m_name
        End Get
        Set(ByVal value As String)
            If Not String.IsNullOrEmpty(value) Then
                m_name = value
            Else
                Throw New ArgumentNullException("value")
            End If
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the schedule rule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The schedule rule.</returns>
    Public Property Rule() As String
        Get
            Return m_minutes.Text & " " & m_hours.Text & " " & m_days.Text & " " & m_months.Text & " " & m_dayOfWeek.Text
        End Get
        Set(ByVal value As String)
            If Not String.IsNullOrEmpty(value) Then
                With RemoveDuplicateWhiteSpace(value).Split(" "c)
                    If .Length = 5 Then
                        Me.Minutes = .GetValue(0).ToString()
                        Me.Hours = .GetValue(1).ToString()
                        Me.Days = .GetValue(2).ToString()
                        Me.Months = .GetValue(3).ToString()
                        Me.DaysOfWeek = .GetValue(4).ToString()
                    Else
                        Throw New ArgumentException("Schedule rule must have exactly 5 elements (Example: * * * * *).")
                    End If
                End With
            Else
                Throw New ArgumentNullException("value")
            End If
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the 1st element of the 5 element schedule rule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The 1st element of the 5 element schedule rule.</returns>
    Public Property Minutes() As String
        Get
            Return m_minutes.Text
        End Get
        Set(ByVal value As String)
            m_minutes = New Element(RemoveWhiteSpace(value), Element.ElementType.Minute)
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the 2nd element of the 5 element schedule rule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The 2nd element of the 5 element schedule rule.</returns>
    Public Property Hours() As String
        Get
            Return m_hours.Text
        End Get
        Set(ByVal value As String)
            m_hours = New Element(RemoveWhiteSpace(value), Element.ElementType.Hour)
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the 3rd element of the 5 element schedule rule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The 3rd element of the 5 element schedule rule.</returns>
    Public Property Days() As String
        Get
            Return m_days.Text
        End Get
        Set(ByVal value As String)
            m_days = New Element(RemoveWhiteSpace(value), Element.ElementType.Day)
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the 4th element of the 5 element schedule rule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The 4th element of the 5 element schedule rule.</returns>
    Public Property Months() As String
        Get
            Return m_months.Text
        End Get
        Set(ByVal value As String)
            m_months = New Element(RemoveWhiteSpace(value), Element.ElementType.Month)
        End Set
    End Property

    ''' <summary>
    ''' Gets or sets the 5th element of the 5 element schedule rule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The 5th element of the 5 element schedule rule.</returns>
    Public Property DaysOfWeek() As String
        Get
            Return m_dayOfWeek.Text
        End Get
        Set(ByVal value As String)
            m_dayOfWeek = New Element(RemoveWhiteSpace(value), Element.ElementType.DayOfWeek)
        End Set
    End Property

    ''' <summary>
    ''' Gets the current status of the schedule.
    ''' </summary>
    ''' <value></value>
    ''' <returns>The current status of the schedule.</returns>
    Public ReadOnly Property Status() As String
        Get
            With New System.Text.StringBuilder()
                .Append("             Schedule name: ")
                .Append(m_name)
                .Append(Environment.NewLine)
                .Append("             Schedule rule: ")
                .Append(Rule)
                .Append(Environment.NewLine)
                .Append("             Last run time: ")
                .Append(IIf(m_lastRunDateTime = System.DateTime.MinValue, "Never", m_lastRunDateTime))
                .Append(Environment.NewLine)

                Return .ToString()
            End With
        End Get
    End Property

    ''' <summary>
    ''' Checks whether the schedule is due at the present system time.
    ''' </summary>
    ''' <returns>True if the schedule is due at the present system time; otherwise False.</returns>
    Public Function IsDue() As Boolean

        Dim currentDateTime As System.DateTime = System.DateTime.Now
        If m_minutes.Matches(currentDateTime) AndAlso m_hours.Matches(currentDateTime) AndAlso _
                m_days.Matches(currentDateTime) AndAlso m_months.Matches(currentDateTime) AndAlso _
                m_dayOfWeek.Matches(currentDateTime) Then
            m_lastRunDateTime = currentDateTime
            Return True
        Else
            Return False
        End If

    End Function

    Public Overrides Function Equals(ByVal obj As Object) As Boolean

        Dim other As Schedule = TryCast(obj, Schedule)
        If other IsNot Nothing AndAlso other.Name = Me.Name AndAlso other.Rule = Me.Rule Then
            Return True
        Else
            Return False
        End If

    End Function

End Class

File: Source/Libraries/GSF.Core/Scheduling/Schedule.cs
//******************************************************************************************************
//  Schedule.cs - Gbtc
//
//  Copyright © 2012, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  08/01/2006 - Pinal C. Patel
//       Generated original version of source code.
//  09/15/2008 - J. Ritchie Carroll
//       Converted to C#.
//  09/14/2009 - Stephen C. Wills
//       Added new header and license agreement.
//  12/14/2012 - Starlynn Danyelle Gilliam
//       Modified Header.
//
//******************************************************************************************************

using System;
using System.ComponentModel;
using System.Text;

namespace GSF.Scheduling
{
    /// <summary>
    /// Represents a schedule defined using UNIX crontab syntax.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Operators:
    /// </para>
    /// <para>
    /// There are several ways of specifying multiple date/time values in a field:
    /// <list type="bullet">
    /// <item>
    ///     <description>
    ///         The comma (',') operator specifies a list of values, for example: "1,3,4,7,8"
    ///     </description>
    /// </item>
    /// <item>
    ///     <description>
    ///         The dash ('-') operator specifies a range of values, for example: "1-6",
    ///         which is equivalent to "1,2,3,4,5,6"
    ///     </description>
    /// </item>
    /// <item>
    ///     <description>
    ///         The asterisk ('*') operator specifies all possible values for a field.
    ///         For example, an asterisk in the hour time field would be equivalent to
    ///         'every hour' (subject to matching other specified fields).
    ///     </description>
    /// </item>
    /// <item>
    ///     <description>
    ///         The slash ('/') operator (called "step"), which can be used to skip a given
    ///         number of values. For example, "*/3" in the hour time field is equivalent
    ///         to "0,3,6,9,12,15,18,21". So "*" specifies 'every hour' but the "*/3" means
    ///         only those hours divisible by 3.
    ///     </description>
    /// </item>
    /// </list>
    /// </para>
    /// <para>
    /// Fields:
    /// </para>
    /// <para>
    /// <code>
    ///     +---------------- minute (0 - 59)
    ///     |  +------------- hour (0 - 23)
    ///     |  |  +---------- day of month (1 - 31)
    ///     |  |  |  +------- month (1 - 12)
    ///     |  |  |  |  +---- day of week (0 - 7) (Sunday=0 or 7)
    ///     |  |  |  |  |
    ///     *  *  *  *  *
    /// </code>
    /// </para>
    /// <para>
    /// Each of the patterns from the first five fields may be either * (an asterisk), which matches all legal values,
    /// or a list of elements separated by commas. 
    /// </para>
    /// <para>
    /// See <a href="http://en.wikipedia.org/wiki/Cron" target="_blank">http://en.wikipedia.org/wiki/Cron</a> for more information.
    /// </para>
    /// </remarks>
    /// <seealso cref="SchedulePart"/>
    /// <seealso cref="ScheduleManager"/>
    public class Schedule
    {
        #region [ Members ]

        // Fields
        private string m_name;
        private string m_description;
        private bool m_useLocalTime;
        private SchedulePart m_minutePart;
        private SchedulePart m_hourPart;
        private SchedulePart m_dayPart;
        private SchedulePart m_monthPart;
        private SchedulePart m_dayOfWeekPart;
        private DateTime m_lastDueAt;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Initializes a new instance of the <see cref="Schedule"/> class.
        /// </summary>
        public Schedule()
            : this("Schedule" + (++s_instances))
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Schedule"/> class.
        /// </summary>
        /// <param name="name">Name of the schedule.</param>
        /// <remarks>Default <see cref="Rule"/> of '* * * * *' is used.</remarks>
        public Schedule(string name)
            : this(name, "* * * * *")
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Schedule"/> class.
        /// </summary>
        /// <param name="name">Name of the schedule.</param>
        /// <param name="rule">Rule formated in UNIX crontab syntax.</param>
        public Schedule(string name, string rule)
            : this(name, rule, "")
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Schedule"/> class.
        /// </summary>
        /// <param name="name">Name of the schedule.</param>
        /// <param name="rule">Rule formated in UNIX crontab syntax.</param>
        /// <param name="description">Description of defined schedule.</param>
        public Schedule(string name, string rule, string description)
            : this(name, rule, description, false)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Schedule"/> class.
        /// </summary>
        /// <param name="name">Name of the schedule.</param>
        /// <param name="rule">Rule formated in UNIX crontab syntax.</param>
        /// <param name="description">Description of defined schedule.</param>
        /// <param name="useLocalTime">Flag that determines whether to use local time for schedule.</param>
        public Schedule(string name, string rule, string description, bool useLocalTime)
        {
            Name = name;
            Rule = rule;
            Description = description;
            UseLocalTime = useLocalTime;
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets or sets the name of the <see cref="Schedule"/>.
        /// </summary>
        /// <exception cref="ArgumentNullException">The value being assigned is null or empty string.</exception>
        public string Name
        {
            get
            {
                return m_name;
            }
            set
            {
                if (string.IsNullOrEmpty(value))
                    throw new ArgumentNullException(nameof(value));

                m_name = value;
            }
        }

        /// <summary>
        /// Gets or sets the rule of the <see cref="Schedule"/> defined in UNIX crontab syntax.
        /// </summary>
        /// <exception cref="ArgumentNullException">The value being assigned is null or empty string.</exception>
        /// <exception cref="ArgumentException">The number of <see cref="SchedulePart"/> in the rule is not exactly 5.</exception>
        public string Rule
        {
            get
            {
                return m_minutePart.ValueText + " " + m_hourPart.ValueText + " " + m_dayPart.ValueText + " " + m_monthPart.ValueText + " " + m_dayOfWeekPart.ValueText;
            }
            set
            {
                if (string.IsNullOrEmpty(value))
                    throw new ArgumentNullException(nameof(value));

                string[] scheduleParts = value.RemoveDuplicateWhiteSpace().Split(' ');

                if (scheduleParts.Length != 5)
                    throw new ArgumentException("Schedule rule must have exactly 5 parts (Example: * * * * *)");

                m_minutePart = new SchedulePart(scheduleParts[0], DateTimePart.Minute);
                m_hourPart = new SchedulePart(scheduleParts[1], DateTimePart.Hour);
                m_dayPart = new SchedulePart(scheduleParts[2], DateTimePart.Day);
                m_monthPart = new SchedulePart(scheduleParts[3], DateTimePart.Month);
                m_dayOfWeekPart = new SchedulePart(scheduleParts[4], DateTimePart.DayOfWeek);

                // Update the schedule description.
                StringBuilder description = new StringBuilder();

                description.Append(m_minutePart.Description);
                description.Append(", ");
                description.Append(m_hourPart.Description);
                description.Append(", ");
                description.Append(m_dayPart.Description);
                description.Append(", ");
                description.Append(m_monthPart.Description);
                description.Append(", ");
                description.Append(m_dayOfWeekPart.Description);

                m_description = description.ToString();
            }
        }

        /// <summary>
        /// Gets or sets a description of the <see cref="Schedule"/>.
        /// </summary>
        /// <remarks>A default description is created automatically when the <see cref="Rule"/> is set.</remarks>
        public string Description
        {
            get
            {
                return m_description;
            }
            set
            {
                if (!string.IsNullOrEmpty(value))
                    m_description = value;
            }
        }

        /// <summary>
        /// Gets or sets a flag that determines whether the scheduler uses local time or UTC time for scheduling.
        /// </summary>
        public bool UseLocalTime
        {
            get
            {
                return m_useLocalTime;
            }
            set
            {
                m_useLocalTime = value;
            }
        }

        /// <summary>
        /// Gets the <see cref="SchedulePart"/> of the <see cref="Schedule"/> that represents minute <see cref="DateTimePart"/>.
        /// </summary>
        [Browsable(false)]
        public SchedulePart MinutePart
        {
            get
            {
                return m_minutePart;
            }
        }

        /// <summary>
        /// Gets the <see cref="SchedulePart"/> of the <see cref="Schedule"/> that represents hour <see cref="DateTimePart"/>.
        /// </summary>
        [Browsable(false)]
        public SchedulePart HourPart
        {
            get
            {
                return m_hourPart;
            }
        }

        /// <summary>
        /// Gets the <see cref="SchedulePart"/> of the <see cref="Schedule"/> that represents day of month <see cref="DateTimePart"/>.
        /// </summary>
        [Browsable(false)]
        public SchedulePart DayPart
        {
            get
            {
                return m_dayPart;
            }
        }

        /// <summary>
        /// Gets the <see cref="SchedulePart"/> of the <see cref="Schedule"/> that represents month <see cref="DateTimePart"/>.
        /// </summary>
        [Browsable(false)]
        public SchedulePart MonthPart
        {
            get
            {
                return m_monthPart;
            }
        }

        /// <summary>
        /// Gets the <see cref="SchedulePart"/> of the <see cref="Schedule"/> that represents day of week <see cref="DateTimePart"/>.
        /// </summary>
        [Browsable(false)]
        public SchedulePart DaysOfWeekPart
        {
            get
            {
                return m_dayOfWeekPart;
            }
        }

        /// <summary>
        /// Gets the <see cref="DateTime"/> when the <see cref="Schedule"/> was last due.
        /// </summary>
        [Browsable(false)]
        public DateTime LastDueAt
        {
            get
            {
                return m_lastDueAt;
            }
        }

        /// <summary>
        /// Gets the current status of the <see cref="Schedule"/>.
        /// </summary>
        [Browsable(false)]
        public string Status
        {
            get
            {
                StringBuilder status = new StringBuilder();

                status.Append("             Schedule name: ");
                status.Append(m_name);
                status.AppendLine();
                status.Append("             Schedule rule: ");
                status.Append(Rule);
                status.AppendLine();
                status.Append("             Last run time: ");
                status.Append(m_lastDueAt == DateTime.MinValue ? "Never" : m_lastDueAt.ToString());
                status.AppendLine();

                return status.ToString();
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Checks whether the <see cref="Schedule"/> is due at the present system time.
        /// </summary>
        /// <returns>true if the <see cref="Schedule"/> is due at the present system time; otherwise false.</returns>
        public bool IsDue()
        {
            DateTime currentDateTime = m_useLocalTime ? DateTime.Now : DateTime.UtcNow;

            if (m_minutePart.Matches(currentDateTime) &&
                m_hourPart.Matches(currentDateTime) &&
                m_dayPart.Matches(currentDateTime) &&
                m_monthPart.Matches(currentDateTime) &&
                m_dayOfWeekPart.Matches(currentDateTime))
            {
                m_lastDueAt = currentDateTime;
                return true;
            }

            return false;
        }

        /// <summary>
        /// Determines the nearest timestamp older than
        /// the given timestamp that matches the schedule.
        /// </summary>
        /// <param name="targetDateTime">The target timestamp</param>
        /// <returns>The timestamp prior to the target that the schedule was previously due</returns>
        public DateTime PreviousTimeDue(DateTime targetDateTime)
        {
            if (m_minutePart.Values.Count == 0 ||
                m_hourPart.Values.Count == 0 ||
                m_dayPart.Values.Count == 0 ||
                m_monthPart.Values.Count == 0 ||
                m_dayOfWeekPart.Values.Count == 0)
            {
                return DateTime.MinValue;
            }

            DateTime currentDateTime = ToScheduleTimeZone(targetDateTime);

            bool FindMatchingMonth()
            {
                while (true)
                {
                    if (m_monthPart.Matches(currentDateTime))
                        return true;
                    if (currentDateTime == DateTime.MinValue)
                        return false;
                    currentDateTime = PreviousMonth(currentDateTime);
                }
            }

            bool FindMatchingDay()
            {
                while (true)
                {
                    if (m_dayPart.Matches(currentDateTime) && m_dayOfWeekPart.Matches(currentDateTime))
                        return true;
                    if (currentDateTime.Day == 1)
                        return false;
                    currentDateTime = PreviousDay(currentDateTime);
                }
            }

            bool FindMatchingHour()
            {
                while (true)
                {
                    if (m_hourPart.Matches(currentDateTime))
                        return true;
                    if (currentDateTime.Hour == 0)
                        return false;
                    currentDateTime = PreviousHour(currentDateTime);
                }
            }

            bool FindMatchingMinute()
            {
                while (true)
                {
                    if (m_minutePart.Matches(currentDateTime))
                        return true;
                    if (currentDateTime.Minute == 0)
                        return false;
                    currentDateTime = PreviousMinute(currentDateTime);
                }
            }

            while (currentDateTime > DateTime.MinValue)
            {
                currentDateTime = PreviousMinute(currentDateTime);

                if (!FindMatchingMonth())
                    continue;

                if (!FindMatchingDay())
                    continue;

                if (!FindMatchingHour())
                    continue;

                if (!FindMatchingMinute())
                    continue;

                return currentDateTime;
            }

            return DateTime.MinValue;
        }

        /// <summary>
        /// Determines the nearest timestamp past the
        /// given timestamp that matches the schedule.
        /// </summary>
        /// <param name="targetDateTime">The target timestamp</param>
        /// <returns>The timestamp past the target that the schedule will next be due</returns>
        public DateTime NextTimeDue(DateTime targetDateTime)
        {
            if (m_minutePart.Values.Count == 0 ||
                m_hourPart.Values.Count == 0 ||
                m_dayPart.Values.Count == 0 ||
                m_monthPart.Values.Count == 0 ||
                m_dayOfWeekPart.Values.Count == 0)
            {
                return DateTime.MaxValue;
            }

            DateTime currentDateTime = ToScheduleTimeZone(targetDateTime);

            bool FindMatchingMonth()
            {
                while (true)
                {
                    if (m_monthPart.Matches(currentDateTime))
                        return true;
                    if (currentDateTime == DateTime.MaxValue)
                        return false;
                    currentDateTime = NextMonth(currentDateTime);
                }
            }

            bool FindMatchingDay()
            {
                while (true)
                {
                    if (m_dayPart.Matches(currentDateTime) && m_dayOfWeekPart.Matches(currentDateTime))
                        return true;

                    DateTime nextDay = NextDay(currentDateTime);

                    if (nextDay.Day == 1)
                        return false;

                    currentDateTime = nextDay;
                }
            }

            bool FindMatchingHour()
            {
                while (true)
                {
                    if (m_hourPart.Matches(currentDateTime))
                        return true;

                    DateTime nextHour = NextHour(currentDateTime);

                    if (nextHour.Hour == 0)
                        return false;

                    currentDateTime = nextHour;
                }
            }

            bool FindMatchingMinute()
            {
                while (true)
                {
                    if (m_minutePart.Matches(currentDateTime))
                        return true;

                    DateTime nextMinute = NextMinute(currentDateTime);

                    if (currentDateTime.Minute == 0)
                        return false;

                    currentDateTime = nextMinute;
                }
            }

            while (currentDateTime < DateTime.MaxValue)
            {
                currentDateTime = NextMinute(currentDateTime);

                if (!FindMatchingMonth())
                    continue;

                if (!FindMatchingDay())
                    continue;

                if (!FindMatchingHour())
                    continue;

                if (!FindMatchingMinute())
                    continue;

                return currentDateTime;
            }

            return DateTime.MaxValue;
        }

        /// <summary>
        /// Gets a hash code for the <see cref="Schedule"/>.
        /// </summary>
        /// <returns>An <see cref="Int32"/> based hashcode.</returns>
        public override int GetHashCode()
        {
            return Rule.GetHashCode();
        }

        /// <summary>
        /// Determines whether the specified <see cref="Schedule"/> is equal to the current <see cref="Schedule"/>.
        /// </summary>
        /// <param name="obj">The <see cref="Schedule"/> to compare with the current <see cref="Schedule"/>.</param>
        /// <returns>
        /// true if the specified <see cref="Schedule"/> is equal to the current <see cref="Schedule"/>; otherwise false
        /// </returns>
        public override bool Equals(object obj)
        {
            Schedule other = obj as Schedule;

            return (object)other != null &&
                   this.Name == other.Name &&
                   this.Rule == other.Rule;
        }

        /// <summary>
        /// Gets the string representation of <see cref="Schedule"/>.
        /// </summary>
        /// <returns>String representation of <see cref="Schedule"/>.</returns>
        public override string ToString()
        {
            return m_name;
        }

        private DateTime ToScheduleTimeZone(DateTime dateTime)
        {
            DateTimeKind targetKind = m_useLocalTime
                ? DateTimeKind.Local
                : DateTimeKind.Utc;

            DateTime specifiedDateTime = (dateTime.Kind == DateTimeKind.Unspecified)
                ? DateTime.SpecifyKind(dateTime, targetKind)
                : dateTime;

            return m_useLocalTime
                ? specifiedDateTime.ToLocalTime()
                : specifiedDateTime.ToUniversalTime();
        }

        #endregion

        #region [ Static ]

        // Static Fields
        private static int s_instances;

        // Static Methods

        private static DateTime PreviousInterval(DateTime dt, TimeSpan interval)
        {
            long ticks = dt.Ticks % interval.Ticks;
            DateTime startOfInterval = dt.AddTicks(-ticks);

            if (startOfInterval <= DateTime.MinValue.AddMinutes(1.0D))
                return DateTime.MinValue;

            return startOfInterval.AddMinutes(-1.0D);
        }

        private static DateTime PreviousMonth(DateTime dt)
        {
            int day = dt.Day;
            DateTime startOfMonth = dt.Date.AddDays(-day + 1.0D);
            return PreviousMinute(startOfMonth);
        }

        private static DateTime PreviousDay(DateTime dt) =>
            PreviousInterval(dt, TimeSpan.FromDays(1.0D));

        private static DateTime PreviousHour(DateTime dt) =>
            PreviousInterval(dt, TimeSpan.FromHours(1.0D));

        private static DateTime PreviousMinute(DateTime dt) =>
            PreviousInterval(dt, TimeSpan.FromMinutes(1.0D));

        private static DateTime NextInterval(DateTime dt, TimeSpan interval)
        {
            long ticks = dt.Ticks % interval.Ticks;
            DateTime startOfInterval = dt.AddTicks(-ticks);

            if (startOfInterval >= DateTime.MaxValue.Add(-interval))
                return DateTime.MaxValue;

            return startOfInterval.Add(interval);
        }

        private static DateTime NextMonth(DateTime dt)
        {
            int day = dt.Day;
            DateTime startOfMonth = dt.Date.AddDays(-day + 1.0D);

            if (startOfMonth >= DateTime.MaxValue.AddMonths(-1))
                return DateTime.MaxValue;

            return startOfMonth.AddMonths(1);
        }

        private static DateTime NextDay(DateTime dt) =>
            NextInterval(dt, TimeSpan.FromDays(1.0D));

        private static DateTime NextHour(DateTime dt) =>
            NextInterval(dt, TimeSpan.FromHours(1.0D));

        private static DateTime NextMinute(DateTime dt) =>
            NextInterval(dt, TimeSpan.FromMinutes(1.0D));

        #endregion

    }
}